### DISCUSSION TITLE
Any quick feedback on rdkit-js? Add it here! And let the community upvote!

### DISCUSSION THREAD

**Discussion by MichelML** [rdkit/rdkit-js#150] [General]
Any quick feedback on rdkit-js? Add it here! And let the community upvote!

**Comment by MaxDNG** [comment]
Love it so far!
One thing that would be awesome would be to add support for SMARTS depiction.
Happy to have a look if it's something others want as well.

**Comment by ptosco** [comment]
@MaxDNG The RDKit (and hence `rdkit-js`) has some support for query atom/bond depiction (e.g., `[C,N]`, `-@` and the like). More complex queries will be depicted as a `?`. If you are thinking of something along the lines of [SMARTSViewer](https://smarts.plus/), we are not there yet... :)

**Reply by MaxDNG** [reply]
That's sort of what I had in mind. SMARTSViewer is indeed a great tool but I was thinking something more simple like displaying the query like you have in MarvinJS.
But again, nothing urgent, just saying in passing and happy to contribute if I can.

**Comment by heberleh** [comment]
I think it is really great and helped me a lot. Thanks a lot to everyone involved.

Because it's about javascript, I would say that some features to identify elements in the page (graph?, positions, ids, hashes, etc) would be nice to have. I solved this by parsing an SVG (getting position of atoms and edges, lengths, etc) and then later generating also a Canvas element, so for each molecule it calculates 2 times the drawings.

One struggle is the 'installation', to make it work with other tools like nodejs, that is, to load it with 'import', etc.
A colleague came the other day to me to ask me how I got it working because they were not getting it. (They gave up, actually :s)
I only got it working because someone (you, I think?) helped me with a way to download it using JavaScript.
I imagine the difficulty has to do with the binary files.
But yeah, once I got how to 'hack', I was able to bring it to React, Jupyter and KNIME. But it always downloads the minimal lib when the webpage loads; no offline approach.

**Reply by MichelML** [reply]
@heberleh thanks for the feedback, I think there are a lot of good points here actually. Definitely the example in the README doesn't cover all the JS contexts in which rdkit-js can be loaded - and this should be resolved. I'm pretty sure there are ways to load the lib offline if the wasm module is already somewhere on your machine, but that needs to be investigated too. I will log two issues for offline usage and one for more import examples.

thank you  again!

**Reply by heberleh** [reply]
Yeah, it has also to do with who is not expert of webpack, etc. Many use sobre tool like create-react-app that already has premade scripts that just work with whatever you import.
I mean, when you transpile and create the bundle.js with everything, that binary needs to go with it in the dist/build folder and the path needs to be all correct. That's my understanding but I didn't try to do that. Sometimes I create my own webpack so I can transpile things into other formats. For my cases it's working and it's ok to use the URL/download version.

**Comment by cbouy** [comment]
I've mainly used it for generating depictions and doing substructure search and it's super useful!

Having examples in the documentation is great and user friendly, but ultimately I think it would be nice to have a proper documentation on the functions arguments and return types.

For example, I know that I can compute descriptors for a given molecule, but I don't know which ones are available, or if there's an option to compute only a specific subset of descriptors. It's relatively easy to dig into the source code or play around in the console to find that info but I think it would be beneficial to have these sort of things explicitly written somewhere accessible.

Thanks again to you, greg and contributors for making, maintaining and packaging this!

**Reply by MichelML** [reply]
Thank you @cbouy , definitely I agree on `ultimately I think it would be nice to have a proper documentation on the functions arguments and return types` . This was the idea behind using manually maintained typescript types and built a documentation on top of it with typedoc . 

You can see the current version at https://docs.rdkitjs.com/ , proper introduction to the api and more type definitions and specific things like `if there's an option to compute only a specific subset of descriptors` should come at some point. If you use typescript with vscode, you should have autocomplete suggestions based on the implemented types, see more on usage here https://github.com/rdkit/rdkit-js/tree/master/typescript

cc @adam-of-barot :point_up:

**Reply by adam-of-barot** [reply]
Could we implement something in the code that contains the available descriptors? The best I can think of is an enum, which could be made available in the typescript docs, but I'm unsure how useful that would be.

**Reply by MichelML** [reply]
@adam-of-barot re-reading @cbouy's comments, I think these were more "exploratory" and a general invitation to keep improving the documentation of the API explicitly than "concrete" suggestions. For example, only computing "all" descriptors is currently available https://github.com/rdkit/rdkit/blob/master/Code/MinimalLib/minilib.h#L37 - yet on that point we could explicitly state all the descriptors it returns here as a first step https://docs.rdkitjs.com/interfaces/JSMol.html#get_descriptors . 

If ever get_descriptors support the computation of a subset of descriptors, we'll add an enum as you suggested :+1:

**Comment by swamidass** [comment]
Thanks for the note, and thanks for developing rdkit! 

The biggest challenge we've had is in extracting the SVG coordinates output from rdkit-js. As it is, I am not sure if it's possible to do so.  We need that to be able to shade molecules by ML predictions. See here, for example, what we are trying to do, and are able to do with the c++ version quite easily: https://xenopict.readthedocs.io/en/latest/quickstart.html.

Would you be willing to add the right interface so that we can do the same in javascript? That would be really nice if that is possible.

**Reply by MichelML** [reply]
@ptosco :point_up: , sorry for the delay @swamidass

**Reply by swamidass** [reply]
Still interested in whether or not a solution is possible. Key things are:

1. To be able to get the plotted coordinates on the SVG of all the atoms.
2. Being able to set drawing parameters when generating the SVG.

With those two features it should be possible to do a lot of powerful things that are not really possible with a canvas.

**Comment by sgoodrow-zymergen** [comment]
Expose more of the rdkit draw configurations, such as using different themes/color palettes.
